### DISCUSSION TITLE
$watch outputs an error

### DISCUSSION THREAD

**Discussion by QJan84** [alpinejs/alpine#3261] [1. Help]
Why is this not working with alpinejs@3.10.3?

```
<section 
  x-data="{
    play: false
  }"
  x-init="$watch('play', value => console.log(play))"
>
```

> Alpine Expression Error: Invalid or unexpected token
> Expression: "$watch('play', value => console.log(play))“
> >

This works fine:
```
<section 
  x-data="{
    play: false
  }"
  x-init="console.log('this works')"
>
```

Something is wrong here with $watch?

**Comment by sunsande** [comment]
I created  a [codepen](https://codepen.io/sunsande/pen/JjZNEbE) example.

```html
<div x-data>
  <input x-model="$store.form.text" 
         x-init="$watch('$store.form.text', (val) => { console.log('x-init ok: ' + val)   })" />
</div>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/alpinejs/3.10.3/cdn.js"></script>
<script>
    document.addEventListener("alpine:init", () => {
      Alpine.store("form", {
        init() {
          console.log("init ...") 
          this.$watch('$store.form.text', (val) => { console.log(`store init ok:  ${val}`)   })
        },
        text: 'xxx'
      });
    });
</script>
```
`x-init` works on both `3.10.3` and `3.10.5`.

But the line in the `store:init` function `this.$watch('$store.form.text', (val) => { console.log(`store init ok:  ${val}`)   })`
throws error:
```
TypeError: this.$watch is not a function 
 at https://cdpn.io/cpe/boomboom/index.html?key=index.html-3665d757-2b0a-64c9-2b7a-f5a798e4deb0:39
```
So `this.$watch` does not work in the `store:init` function even though there are a lot examples on the web using `this.$watch`. Could someone explain why does `this.$watch` not working?

**Reply by ekwoka** [reply]
Stores don't have access to all of the magics.

Often watch should be replaced by effect.

**Reply by sunsande** [reply]
Thank you @ekwoka for the quick reply!
Is there a reason why one could not have the watch into the store? It is quite easy to make something like this:
```html
<div x-data x-init="$store.someStore.$watch = $watch">
```
and then it is available in the store and works fine. But why not make it available directly out of the box? Is there some scenario where it may cause a problem?

**Reply by ekwoka** [reply]
I'm not sure what the initial reason is (none of the Magics are provided to stores and I know some require elements). May have just been implementation complexity in the system.

Another can be that it's generally better to use `Alpine.effect` when doing more programatic stuff

The example you show can just be

```js
Alpine.effect(()=>console.log(`store init ok:  ${this.text}`))
```

**Reply by SimoTod** [reply]
All the magics are bound to an html element so any side effects get automatically cleared if the element gets removed from the DOM. The naked store is not bound to any specific element so no magics. What you are doing in your hack is to set the element $watch into the store (which element should do that? What if you have many? If it's just one, probably you don't need a store but Alpine.data, etc). As mentioned, Alpine.effect is not element bound and available everywhere.

**Comment by sunsande** [comment]
@SimoTod  Thank you for the comments!
I do not know any alpine.js implementation details, so I could not comment on that. The documentation does not help a lot in this direction either.
> All the magics are bound to an html element so ...

and the documentation says:
>You can "watch" a component property using the $watch magic method.

But if you look from alpine.js user's perspective the following statement works perfect:
```javascript
$watch('$store.someStore.someProperty', () => { ... })
```
And it does not actually matter from what element we get the reference to the `$watch`. It does not matter if there are many - one is enough to take the reference to the `$watch` and then we can use it `N` times:
```javascript
$watch('$store.someStore.property1', () => { ... })
$watch('$store.someOtherStore.property2', () => { ... })
...
$watch('$store.someMoreOtherStore.propertyN', () => { ... })
```
As I said , I do not know if it is "right" and if there are side effects etc. ... but it is really useful and it is not the same as the `effect` function - there are at least 2 differences: 1) with `$watch` one could get old/new value 2) with $watch one could specify a separate function for every `watched` property (whereas `effect` is only one function). 

And assuming all that it is easy to see why would one ask: why `$watch` would not work directly in a store without any tricks. It is actually working perfectly - the only trick needed is to provide the reference to the `$watch` inside the store - then one can use it as often as needed.  (I do not know what happens if the DOM element, which provided the reference to the `$watch` gets deleted? The documentation could do a better job explaining that ... any help/additional information is great appreciated!)

**Reply by SimoTod** [reply]
I agree, the documentation could be better. Nevertheless, magics only work in components. $watch internally uses an effect so it should be possible to implement the same features.

When you remove the element, the watch callback won't be trigger anymore when the watched property changes.

The source code is not huge if you are interested in knowing the details about the implementation. Unfortunately it's a lot to cover in a thread and I'm just a user like you, so I can't answer questions about design choices.
